text/3556-re-export-symbols.md

```diff
@@ -0,0 +1,131 @@
+- Feature Name: `re_export_symbols`
+- Start Date: 2024-01-05
+- RFC PR: [rust-lang/rfcs#3556](https://github.com/rust-lang/rfcs/pull/3556)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow users to build a cdylib rust crate which both:
+
+ - links against a pre-built staticlib
+ - re-exports specific symbols from that staticlib
+
+# Motivation
+[motivation]: #motivation
+
+Currently the Rust compiler attempts to restrict visibility of symbols not intended for export. This is good practice when building cdylibs (`.so` on Linux) as it produces much smaller linked output.
+
+When building a cdylib, one will typically link to native dependencies either as 1) a cdylib or 2) a staticlib. The latter case can be preferable when trying to make the final cdylib product 'freestanding', with minimal dependencies on the host system.
+
+When linking a staticlib into a cdylib crate, there are some cases where it is useful to re-export symbols. There is extensive dicussion from [people wanting to do this](https://github.com/rust-lang/rfcs/issues/2771) (with a note that the Lang team discussed this in 2017!), as well as another issue from people attempting to [use `+whole-archive`](https://github.com/rust-lang/rust/issues/110624) to re-export _all_ symbols from a linked staticlib.
+
+However, the *only* supported way to re-export a symbol today is manually write wrappers around the staticlib symbols, re-passing the arguments down. This is:
+ - inconvenient - re-exporting as the same name requires putting the `extern` block and wrapper function in different modules due to name collision
+ - error-prone - the user has to maintain the wrapper function with the correct arguments
+ - sometimes impossible - variadic functions are [not stable yet](https://github.com/rust-lang/rust/issues/44930)
+
+This tightly-scoped RFC adds support for re-exporting explicitly listed symbols from a linked staticlib in an extern block to the final produced cdylib. It does *not* add 'wildcard' support to support re-exporting all symbols from a staticlib, nor does it cover other staticlib/cdylib combinations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When producing a cdylib from a Rust crate and linking to a staticlib, adding `#[no_mangle]` and `pub` to an item in an extern block in the top level crate as follows:
+
+```
+#[no_mangle]
+pub fn foo();
```

Comment:
Can you put this example inside an extern block to make it more obvious?

Reply:
Done.